### PR TITLE
CXP-558: Dispatch a BulkEvent for product model

### PR DESCRIPTION
**Pull request by pierrejolly**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

**Review by pchasle — APPROVED** [review]

File: src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/BusinessEvent/DispatchProductModelCreatedAndUpdatedEventSubscriber.php [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+            StorageEvents::POST_SAVE_ALL => 'dispatchBufferedProductModelEvents',
+        ];
+    }
+
+    public function createAndDispatchProductModelEvents(GenericEvent $postSaveEvent): void
+    {
+        /** @var ProductModelInterface */
+        $productModel = $postSaveEvent->getSubject();
+        if (false === $productModel instanceof ProductModelInterface) {
+            return;
```
Comment (tomglvng) [inline src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/BusinessEvent/DispatchProductModelCreatedAndUpdatedEventSubscriber.php:52]: Do we need to throw an exception here in CXP-563 ? If yes, please add a "TODO" comment.

File: src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/BusinessEvent/DispatchProductModelCreatedAndUpdatedEventSubscriber.php [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        }
+
+        if (null === $user = $this->getUser()) {
+            return;
```
Comment (tomglvng) [inline src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/BusinessEvent/DispatchProductModelCreatedAndUpdatedEventSubscriber.php:56]: Same question here.

File: src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/BusinessEvent/DispatchProductModelRemovedEventSubscriber.php [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+            StorageEvents::POST_REMOVE => 'createAndDispatchProductModelEvents',
+        ];
+    }
+
+    public function createAndDispatchProductModelEvents(GenericEvent $postSaveEvent): void
+    {
+        /** @var ProductModelInterface */
+        $productModel = $postSaveEvent->getSubject();
+        if (false === $productModel instanceof ProductModelInterface) {
+            return;
```
Comment (tomglvng) [inline src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/BusinessEvent/DispatchProductModelRemovedEventSubscriber.php:44]: TODO in CXP-563 ?

**Review by tomglvng — APPROVED** [review]

File: src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/BusinessEvent/DispatchProductModelCreatedAndUpdatedEventSubscriber.php (hunk 1 above)
Comment (pchasle) [reply src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/BusinessEvent/DispatchProductModelCreatedAndUpdatedEventSubscriber.php:52]: No we don't need to :) this is a standard behavior of any event subscribers on a generic event (and post save event subscriber).
We need to validate the type of the "subject" object to not handle Attributes, simple Product, etc...

**Pull request merged**